src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -372,7 +372,97 @@ protected void InitializeReferences()
             try
             {
-                return runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                runner.InitializeTimeout(internalMatchTimeout);
+                runner.runtext = input;
+                ReadOnlySpan<char> span = input.AsSpan(beginning, length);
+                runner.InitializeForScan(this, span, startat - beginning, quick);
+
+                int stoppos = RightToLeft ? 0 : span.Length;
+
+                // If previous match was empty or failed, advance by one before matching.
+                if (prevlen == 0)
+                {
+                    if (runner.runtextstart == stoppos)
+                    {
+                        return RegularExpressions.Match.Empty;
+                    }
+
+                    runner.runtextpos += RightToLeft ? -1 : 1;
+                }
+
+                return InternalPerformScan(quick, input, beginning, runner, span, returnNullIfQuick: true);
+            }
+            finally
+            {
+                runner.runtext = null; // drop reference to text to avoid keeping it alive in a cache.
+                _runner = runner;
+            }
+        }
+
+        private static Match? InternalPerformScan(bool quick, string input, int beginning, RegexRunner runner, ReadOnlySpan<char> span, bool returnNullIfQuick)
+        {
+            runner.Scan(span);
+
+            Match? match = runner.runmatch;
+            Debug.Assert(match is not null);
+
+            // If we got a match, set runmatch to null if quick is true.
+            if (match.FoundMatch)
+            {
+                if (!quick)
+                {
+                    match.Text = input; // We need to save the input into the match object which will be returned.
+                }
+
+                if (quick && returnNullIfQuick)
+                {
+                    return null;
+                }
+
+                if (!quick)
+                {
+                    runner.runmatch = null;
+                }
+
+                match.Tidy(runner.runtextpos);
+
+                // If the passed in beginning was not 0 then we need to adjust the offests on the match object.
+                if (beginning != 0)
+                {
+                    match.AddBeginningToIndex(beginning);
+                }
+
+                return match;
+            }
+
+            if (!quick)
+            {
+                runner.runmatch = null;
+            }
+            else
+            {
+                match.Text = null;
+            }
+
+            return RegularExpressions.Match.Empty;
+        }
+
+        internal Match? Run(ReadOnlySpan<char> input, int startat)
+        {
+            if ((uint)startat > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startat, ExceptionResource.BeginIndexNotNegative);
```

Comment:
Code coverage shows this is never hit.  We should ensure there are tests that do so, or if it's unreachable understand why and delete it.